### DISCUSSION TITLE
Selecting mark character separately in the cluster

### DISCUSSION THREAD

**Discussion by DeependraKhetwal** [harfbuzz/harfbuzz#4575] [General] [answered]
Hello I am trying to implement a feature where the users can select base characters differently and mark characters differently within a cluster and can place cursor in between the cluster ( like shown in example below , the text is RTL hebrew). Is there some information which i can get from harfbuzz which can help me achieve this. I am currently using the default cluster level 0 but i am not able to ascertain from this how to divide the visual area of cluster in individual glyphs. If i change to cluster level 1 then the diacritic becomes separate cluster and since i am selecting visually the selection applies only on base character . Any ideas here about how we may achieve this?

**Comment by DeependraKhetwal** [comment]
![mark_character_selected](https://github.com/harfbuzz/harfbuzz/assets/122281418/abb36398-e36a-4adb-bd60-87333db48323) mark character selected

**Comment by DeependraKhetwal** [comment]
![base_letter_selected](https://github.com/harfbuzz/harfbuzz/assets/122281418/25ee342c-3e15-4974-b50d-fab28b52391f)
base character selected

**Reply by DeependraKhetwal** [reply]
@behdad / @khaledhosny  : Any ideas how we can support visual selection of mark character within a cluster ? And Is there a way to identify within a cluster which unicode characters map to which mark glyphs exactly? One of our customers want to select mark characters separately and apply styling,color etc on them.

**Reply by behdad** [reply, marked as answer]
Depends on the font. If they font does not include the mark as a separate glyph, then you can't do what you want. If it does though, you can get to that by using `hb_buffer_set_cluster_level(buffer, HB_BUFFER_CLUSTER_LEVEL_MONOTONE_CHARACTERS)`.

**Reply by DeependraKhetwal** [reply]
@behdad Actually we want to be able to give users the option to visually select the mark characters as part of the cluster containing the base character( something like the snaps i attached above) . So if we change to HB_BUFFER_CLUSTER_LEVEL_MONOTONE_CHARACTERS then selecting visually becomes a challenge because when user selects a part of base character visually using mouse/keyboard then they are supposed to select mark character but since with HB_BUFFER_CLUSTER_LEVEL_MONOTONE_CHARACTERS  the base character is in a separate cluster and mark characters are in separate cluster it becomes a challenge. is there is way to identify with cluster level 0 (HB_BUFFER_CLUSTER_LEVEL_MONOTONE_GRAPHEMES) which glyphs of clusters are base and which are mark and their exact mapping to input unicodes?

**Reply by behdad** [reply]
HarfBuzz under the default setting gives you clusters that are fairly safe to use as a selection unit. But it's not designed to be used that way.

You need to implement the Unicode graphemes yourself if you want to achieve what you want:
https://unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries

Again, as I said, if the font is designed to have the mark and base as one glyph, there is no way to select the mark separately.

**Reply by DeependraKhetwal** [reply]
@behdad If I am not wrong grapheme cluster boundaries will return the base and mark character together as they are part of one grapheme cluster. So i think I would not be able to separate them using above.

**Reply by behdad** [reply]
You would use Unicode grapheme clusters for selection and mouse movement, but HarfBuzz clusters under `MONOTONE_CHARACTERS` for coloring the marks.
